[TensorExpr] Add a method for sanitizing Var and Buf names in Stmt.

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #65010

This pass ensures all names are legal and not-duplicated.

Fixes #52727.

Differential Revision: [D30939717](https://our.internmc.facebook.com/intern/diff/D30939717)

Commit:
[TensorExpr] Add a method for sanitizing Var and Buf names in Stmt.

This pass ensures all names are legal and not-duplicated.

Fixes #52727.

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30939717).

navahgar commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -6698,5 +6698,68 @@ TEST(LoopNest, compressMultipleBuffers) {
+# CHECK-NEXT:                        for (int l1 = 0; l1 < N; l1++) {
+# CHECK-NEXT:                          v_X__[i, j, k, l, m, n, o, p, i1, j1, k1, l1] = ((i + j) + j1) + 1;
+# CHECK:  v_X___1 = int(0);
+# CHECK-NEXT:  for (int i_2 = 0; i_2 < i_1; i_2++) {
```
Why is this `i_2` and not `i_1`, since `i_1` doesn't seem to be "taken" here?

ZolotukhinM replied:
`i_1` is (intentionally) used as a stop value for the very first loop, so we have to go with `i_2` here.

navahgar replied:
Oh yeah, okay. I missed that.

navahgar commented on torch/csrc/jit/tensorexpr/loopnest.h:
```diff
@@ -109,6 +109,17 @@ class TORCH_API LoopNest {
+  // The pass assigns predefined names for loop index variables
+  // (i,j,k,l,m,n,o,p,i1,j1,k1,...) and ensures these names are not conflicting
+  // anywhere. It also removes duplicates from other Buf nad Var names as well
+  // as replaces illegal characters in them with underscores.
```
Can you also add a comment to say how the index variables are named uniquely?
Something like `<index_var_name><level / number of index vars>_...`

ZolotukhinM replied:
That's mentioned in "(i,j,k,l,m,n,o,p,i1,j1,k1,...)". I didn't want to be overly specific because it's not critical as long as the names are legal and unique.

navahgar commented on torch/csrc/jit/tensorexpr/kernel.cpp:
```diff
@@ -2623,6 +2623,7 @@ static void parallelizeOuterLoops(LoopNest& l, Bufs&& bufs) {
 
 StmtPtr TensorExprKernel::transformLoops(BackendType backendType, StmtPtr st) {
   torch::jit::tensorexpr::LoopNest l(st, bufOutputs_);
+  LoopNest::sanitizeNames(l.root_stmt());
```
One other alternative is to call `sanitizeNames` from the `LoopNest` constructor. That way we can always ensure all stmts in `LoopNest` adhere to this.

ZolotukhinM replied:
Right, we can do that too, I just hoped that we at some point kill LoopNest entirely :D

navahgar reviewed (commented):
Mostly LGTM.

This name sanitization is done at the start of the loop transformations. Should we do it at later stages as well? Technically, loop transformations may result in the IR not following this.

ZolotukhinM commented:
> This name sanitization is done at the start of the loop transformations. Should we do it at later stages as well? Technically, loop transformations may result in the IR not following this.

I thought about this as well, but decided to not do the second sanitization because if we, let's say, split one of the loop with i, j, k indices, I think it's better to end up with i, j_outer, j_inner, k than i,j,k,l. If our passes behave correctly, then the sanitation afterwards should not be needed, even though as you noticed we might be using not-first names anymore.

Commit:
Update on "[TensorExpr] Add a method for sanitizing Var and Buf names in Stmt."


This pass ensures all names are legal and not-duplicated.

Fixes #52727.

Differential Revision: [D30939717](https://our.internmc.facebook.com/intern/diff/D30939717)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30939717).
